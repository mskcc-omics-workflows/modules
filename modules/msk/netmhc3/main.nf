process NETMHC3 {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'docker://mskcc/netmhctools:1.1.0':
        'docker.io/mskcc/netmhctools:1.1.0' }"

    input:
    tuple val(meta),  path(inputFasta), path(inputSVFasta, arity: '0..*'), val(hlaString), val(inputType)

    output:
    tuple val(output_meta),       path("*.xls"),               emit: xls
    tuple val(output_meta),       path("*.netmhc.output"),     emit: netmhcoutput
    tuple val(output_meta),       path("*.hla_*.txt"),         emit: netmhc_hla_files
    path "versions.yml",                                       emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def hla = hlaString.trim()
    output_meta = meta.clone()
    output_meta.typeMut = inputType == "MUT" ? true : false
    output_meta.fromStab = false
    output_meta.fromPan = false
    def NETMHC_VERSION = "3.4"
    def tmpDir = "netmhc-tmp"
    def tmpDirFullPath = "\$PWD/${tmpDir}/"  // must set full path to tmp directories for netMHC and netMHCpan to work; for some reason doesn't work with /scratch, so putting them in the process workspace

    """
    export TMPDIR=${tmpDirFullPath}
    mkdir -p ${tmpDir}
    chmod 777 ${tmpDir}

    HLA_ACCEPTED=\$(trim_hla.py --hla ${hla})

    cat ${inputSVFasta} >> ${inputFasta}

    /usr/local/bin/netMHC-3.4/netMHC \
    -a \$HLA_ACCEPTED \
    -s \
    -l 9 \
    --xls=${prefix}.${inputType}.xls \
    ${inputFasta} > ${prefix}.${inputType}.netmhc.output

    mv hla_accepted.txt ${prefix}.hla_accepted.txt
    mv hla_rejected.txt ${prefix}.hla_rejected.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        netmhc: v${NETMHC_VERSION}
    END_VERSIONS

    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def NETMHC_VERSION = "3.4"
    output_meta = meta.clone()
    output_meta.typeMut = inputType == "MUT" ? true : false
    output_meta.fromStab = false
    output_meta.typePan = false
    """
    touch ${prefix}.MUT.netmhc.output
    touch ${prefix}.MUT.xls
    touch ${prefix}.hla_accepted.txt
    touch ${prefix}.hla_rejected.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        netmhc: v${NETMHC_VERSION}
    END_VERSIONS
    """
}
