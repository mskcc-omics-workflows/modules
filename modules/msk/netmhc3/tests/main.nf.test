nextflow_process {

    name "Test Process NETMHC3"
    script "../main.nf"
    process "NETMHC3"

    tag "modules"
    tag "modules_nfcore"
    tag "netmhc3"
    tag "modules_msk"

    test("netmhc3 - MUT,SV - xls,output,fa") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(file(params.test_data_mskcc['neoantigen']['MUT_sequence_fa']), checkIfExists: true),
                    file(file(params.test_data_mskcc['neoantigen']['svfa']), checkIfExists: true),
                    "HLA-A24:02,HLA-A24:02,HLA-B39:01,HLA-B39:01,HLA-C07:01,HLA-C06:02",
                    "MUT"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.netmhcoutput[0][0],
                    file(process.out.xls[0][1]).name,
                    file(process.out.netmhcoutput[0][1]).name
                    ).match()
                }
            )
        }

    }

    test("netmhc3 - WT,SV - xls,output,fa") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(file(params.test_data_mskcc['neoantigen']['WT_sequence_fa']), checkIfExists: true),
                    file(file(params.test_data_mskcc['neoantigen']['wtsvfa']), checkIfExists: true),
                    "HLA-A24:02,HLA-A24:02,HLA-B39:01,HLA-B39:01,HLA-C07:01,HLA-C06:02",
                    "WT"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.netmhcoutput[0][0],
                    file(process.out.xls[0][1]).name,
                    file(process.out.netmhcoutput[0][1]).name
                    ).match()
                }
            )
        }

    }

    test("netmhc3 - MUT - xls,output,fa") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(file(params.test_data_mskcc['neoantigen']['MUT_sequence_fa']), checkIfExists: true),
                    [],
                    "HLA-A24:02,HLA-A24:02,HLA-B39:01,HLA-B39:01,HLA-C07:01,HLA-C06:02",
                    "MUT"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.netmhcoutput[0][0],
                    file(process.out.xls[0][1]).name,
                    file(process.out.netmhcoutput[0][1]).name
                    ).match()
                }
            )
        }

    }

    test("netmhc3 - WT - xls,output,fa") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(file(params.test_data_mskcc['neoantigen']['WT_sequence_fa']), checkIfExists: true),
                    [],
                    "HLA-A24:02,HLA-A24:02,HLA-B39:01,HLA-B39:01,HLA-C07:01,HLA-C06:02",
                    "WT"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.netmhcoutput[0][0],
                    file(process.out.xls[0][1]).name,
                    file(process.out.netmhcoutput[0][1]).name
                    ).match()
                }
            )
        }

    }



    test("netmhc3 - xls,output,fa - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file('MUT_sequence_fa'),
                    file('svfa'),
                    "HLA",
                    "MUT"
                    ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.versions,
                    process.out.netmhcoutput[0][0],
                    file(process.out.xls[0][1]).name,
                    file(process.out.netmhcoutput[0][1]).name
                    ).match()
                }
            )
        }

    }

}
