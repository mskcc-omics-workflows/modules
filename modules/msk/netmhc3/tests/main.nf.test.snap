{
    "netmhc3 - MUT,SV - xls,output,fa": {
        "content": [
            [
                "versions.yml:md5,41df6059e67545fe2fe2734514557bb7"
            ],
            {
                "id": "test",
                "single_end": false,
                "typeMut": true,
                "fromStab": false,
                "fromPan": false
            },
            "test.MUT.xls",
            "test.MUT.netmhc.output"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T12:08:28.261935482"
    },
    "netmhc3 - xls,output,fa - stub": {
        "content": [
            [
                "versions.yml:md5,41df6059e67545fe2fe2734514557bb7"
            ],
            {
                "id": "test",
                "single_end": false,
                "typeMut": true,
                "fromStab": false,
                "typePan": false
            },
            "test.MUT.xls",
            "test.MUT.netmhc.output"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T12:09:30.886623638"
    },
    "netmhc3 - WT,SV - xls,output,fa": {
        "content": [
            [
                "versions.yml:md5,41df6059e67545fe2fe2734514557bb7"
            ],
            {
                "id": "test",
                "single_end": false,
                "typeMut": false,
                "fromStab": false,
                "fromPan": false
            },
            "test.WT.xls",
            "test.WT.netmhc.output"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T12:08:47.429513751"
    },
    "netmhc3 - WT - xls,output,fa": {
        "content": [
            [
                "versions.yml:md5,41df6059e67545fe2fe2734514557bb7"
            ],
            {
                "id": "test",
                "single_end": false,
                "typeMut": false,
                "fromStab": false,
                "fromPan": false
            },
            "test.WT.xls",
            "test.WT.netmhc.output"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T12:09:21.112597962"
    },
    "netmhc3 - MUT - xls,output,fa": {
        "content": [
            [
                "versions.yml:md5,41df6059e67545fe2fe2734514557bb7"
            ],
            {
                "id": "test",
                "single_end": false,
                "typeMut": true,
                "fromStab": false,
                "fromPan": false
            },
            "test.MUT.xls",
            "test.MUT.netmhc.output"
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-12-17T12:09:04.393123362"
    }
}